File: src/agent/codex-appserver/app-server-client.ts
import { spawn, type ChildProcessWithoutNullStreams } from 'node:child_process';
import { log } from '../../core/logger';
import type { ServerNotification } from './protocol';

/** Simple async queue: push() from the reader, async-iterate from consumers. */
class AsyncQueue<T> {
  private items: T[] = [];
  private waiters: ((v: IteratorResult<T>) => void)[] = [];
  private closed = false;

  push(item: T): void {
    const w = this.waiters.shift();
    if (w) w({ value: item, done: false });
    else this.items.push(item);
  }

  close(): void {
    this.closed = true;
    while (this.waiters.length) this.waiters.shift()!({ value: undefined as never, done: true });
  }

  async *[Symbol.asyncIterator](): AsyncGenerator<T> {
    while (true) {
      if (this.items.length) {
        yield this.items.shift()!;
        continue;
      }
      if (this.closed) return;
      const next = await new Promise<IteratorResult<T>>((resolve) => this.waiters.push(resolve));
      if (next.done) return;
      yield next.value;
    }
  }
}

interface Pending {
  resolve: (v: unknown) => void;
  reject: (e: Error) => void;
}

export interface AppServerClientOptions {
  bin: string;
  cwd: string;
  env?: Record<string, string>;
  clientName?: string;
}

/**
 * One `codex app-server --listen stdio://` child process, speaking JSON-RPC 2.0
 * over newline-delimited JSON. One client = one thread/session (per design:
 * a process per session for crash isolation).
 */
export class AppServerClient {
  private child: ChildProcessWithoutNullStreams | null = null;
  private buf = '';
  private nextId = 0;
  private readonly pending = new Map<number, Pending>();
  private readonly notifications = new AsyncQueue<ServerNotification>();
  private closed = false;

  constructor(private readonly opts: AppServerClientOptions) {}

  get pid(): number | undefined {
    return this.child?.pid;
  }

  /** spawn + initialize handshake. Throws if spawn/handshake fails. */
  async connect(): Promise<void> {
    const child = spawn(this.opts.bin, ['app-server', '--listen', 'stdio://'], {
      cwd: this.opts.cwd,
      env: { ...process.env, ...this.opts.env, FEISHU_CODEX_BRIDGE: '1' },
      stdio: ['pipe', 'pipe', 'pipe'],
    });
    this.child = child;
    log.info('agent', 'spawn', { pid: child.pid ?? null, cwd: this.opts.cwd });

    child.stdout.on('data', (d: Buffer) => this.onStdout(d));
    child.stderr.on('data', (d: Buffer) => {
      const line = d.toString('utf8').trim();
      if (line) log.warn('agent', 'stderr', { line: line.slice(0, 200) });
    });
    child.on('exit', (code, signal) => {
      log.info('agent', 'exit', { pid: child.pid ?? null, code, signal });
      this.failAllPending(new Error(`app-server exited (code=${code} signal=${signal})`));
      this.notifications.close();
    });
    child.on('error', (err) => this.failAllPending(err));

    await this.request('initialize', {
      clientInfo: { name: this.opts.clientName ?? 'feishu-codex-bridge', version: '0.0.1' },
      capabilities: null,
    });
    this.notify('initialized');
  }

  request<T = unknown>(method: string, params?: unknown): Promise<T> {
    if (this.closed || !this.child) return Promise.reject(new Error('app-server client closed'));
    const id = ++this.nextId;
    const payload = `${JSON.stringify({ jsonrpc: '2.0', id, method, params: params ?? {} })}\n`;
    return new Promise<T>((resolve, reject) => {
      this.pending.set(id, { resolve: resolve as (v: unknown) => void, reject });
      this.child!.stdin.write(payload, (err) => {
        if (err) {
          this.pending.delete(id);
          reject(err);
        }
      });
    });
  }

  notify(method: string, params?: unknown): void {
    if (this.closed || !this.child) return;
    this.child.stdin.write(`${JSON.stringify({ jsonrpc: '2.0', method, params: params ?? {} })}\n`);
  }

  /** async-iterate server notifications (closes when the process exits). */
  stream(): AsyncIterable<ServerNotification> {
    return this.notifications;
  }

  async close(graceMs = 4000): Promise<void> {
    if (this.closed) return;
    this.closed = true;
    const child = this.child;
    if (!child || child.exitCode !== null) return;
    child.kill('SIGTERM');
    await new Promise<void>((resolve) => {
      const t = setTimeout(() => {
        if (child.exitCode === null) child.kill('SIGKILL');
        resolve();
      }, graceMs);
      child.once('exit', () => {
        clearTimeout(t);
        resolve();
      });
    });
  }

  private onStdout(d: Buffer): void {
    this.buf += d.toString('utf8');
    let nl: number;
    while ((nl = this.buf.indexOf('\n')) !== -1) {
      const line = this.buf.slice(0, nl);
      this.buf = this.buf.slice(nl + 1);
      if (!line.trim()) continue;
      this.handleLine(line);
    }
  }

  private handleLine(line: string): void {
    let msg: Record<string, unknown>;
    try {
      msg = JSON.parse(line);
    } catch {
      log.warn('agent', 'nonjson', { line: line.slice(0, 120) });
      return;
    }

    // response to one of our requests
    if (typeof msg.id === 'number' && (('result' in msg) || ('error' in msg)) && !('method' in msg)) {
      const p = this.pending.get(msg.id);
      if (!p) return;
      this.pending.delete(msg.id);
      if ('error' in msg && msg.error) {
        const e = msg.error as { message?: string };
        p.reject(new Error(e.message ?? 'JSON-RPC error'));
      } else {
        p.resolve(msg.result);
      }
      return;
    }

    // server-initiated request (e.g. approval). With approvalPolicy:never these
    // shouldn't arrive, but reply method-not-found so the server never blocks.
    if (typeof msg.id === 'number' && typeof msg.method === 'string') {
      this.child?.stdin.write(
        `${JSON.stringify({ jsonrpc: '2.0', id: msg.id, error: { code: -32601, message: 'not handled' } })}\n`,
      );
      return;
    }

    // notification
    if (typeof msg.method === 'string') {
      this.notifications.push(msg as unknown as ServerNotification);
    }
  }

  private failAllPending(err: Error): void {
    for (const p of this.pending.values()) p.reject(err);
    this.pending.clear();
  }
}

File: internal/agent/codex/appserver_client.go
package codex

import (
	"bufio"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"os/exec"
	"sync"
	"time"

	"github.com/modelzen/feishu-codex-bridge/internal/core"
)

// appserver_client.go —— 一个 `codex app-server --listen stdio://` 子进程，
// 走 JSON-RPC 2.0 over newline-delimited JSON（对齐 TS app-server-client）。
//
// 一个 client = 一个 thread/session（进程级崩溃隔离）。Request 按 id 匹配 response；
// 通知推 chan；进程退出 → 标记 exited + failAllPending + 关闭通知 chan。
//
// JsonRpcError 区分「应用层 RPC error（进程健康）」与「传输层失败/超时（进程坏了）」，
// client-pool 据此决定是否杀共享进程。

// JsonRpcError 应用层 JSON-RPC error（进程健康，只是这次 RPC 失败）。
type JsonRpcError struct{ Msg string }

func (e *JsonRpcError) Error() string { return e.Msg }

// AsJsonRpcError 判定是否 JsonRpcError。
func AsJsonRpcError(err error) bool {
	var j *JsonRpcError
	return errors.As(err, &j)
}

// AppServerClientOptions 构造参数。
type AppServerClientOptions struct {
	Bin        string
	Cwd        string
	Env        map[string]string
	ClientName string
}

// AppServerClient 一个 codex app-server 子进程客户端。
type AppServerClient struct {
	bin        string
	cwd        string
	env        map[string]string
	clientName string

	cmd   *exec.Cmd
	stdin io.WriteCloser

	mu       sync.Mutex
	nextID   int
	pending  map[int]chan rpcResult
	notifyCh chan ServerNotification
	closed   bool
	exited   bool

	exitMu sync.Once
}

type rpcResult struct {
	result json.RawMessage
	err    error
}

// NewAppServerClient 构造（未连接）。
func NewAppServerClient(opts AppServerClientOptions) *AppServerClient {
	name := opts.ClientName
	if name == "" {
		name = "feishu-codex-bridge"
	}
	return &AppServerClient{
		bin:        opts.Bin,
		cwd:        opts.Cwd,
		env:        opts.Env,
		clientName: name,
		pending:    map[int]chan rpcResult{},
		notifyCh:   make(chan ServerNotification, 1024),
	}
}

// Pid 返回子进程 pid（未连接为 0）。
func (c *AppServerClient) Pid() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.cmd != nil && c.cmd.Process != nil {
		return c.cmd.Process.Pid
	}
	return 0
}

// Exited 进程是否已退出（崩了/close）。调用方据此驱逐 thread，让 resume 兜底接管。
func (c *AppServerClient) Exited() bool {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.exited
}

// Connect spawn + initialize 握手。
func (c *AppServerClient) Connect(ctx context.Context) error {
	cmd := exec.Command(c.bin, "app-server", "--listen", "stdio://")
	cmd.Dir = c.cwd
	cmd.Env = append(os.Environ(), "FEISHU_CODEX_BRIDGE=1")
	for k, v := range c.env {
		cmd.Env = append(cmd.Env, k+"="+v)
	}
	setChildSysProcAttr(cmd)

	stdin, err := cmd.StdinPipe()
	if err != nil {
		return err
	}
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return err
	}
	cmd.Stderr = stderrLogger{}
	if err := cmd.Start(); err != nil {
		return fmt.Errorf("spawn codex app-server: %w", err)
	}
	c.cmd = cmd
	c.stdin = stdin
	core.Info(ctx, "agent", "spawn", fmt.Sprintf("codex app-server pid=%d cwd=%s", cmd.Process.Pid, c.cwd))

	go c.readLoop(stdout)
	go func() {
		_ = cmd.Wait()
		c.handleExit()
	}()

	// initialize 握手：experimentalApi 必须开（否则 goal RPC 被拒）。
	_, err = c.Request(ctx, "initialize", map[string]any{
		"clientInfo":   map[string]any{"name": c.clientName, "version": "0.0.1"},
		"capabilities": map[string]any{"experimentalApi": true, "requestAttestation": false},
	})
	if err != nil {
		_ = c.Close(0)
		return err
	}
	c.Notify("initialized", nil)
	return nil
}

// Request 发 JSON-RPC 请求，等响应（受 ctx 超时控制）。
func (c *AppServerClient) Request(ctx context.Context, method string, params any) (json.RawMessage, error) {
	c.mu.Lock()
	if c.closed {
		c.mu.Unlock()
		return nil, errors.New("app-server client closed")
	}
	c.nextID++
	id := c.nextID
	ch := make(chan rpcResult, 1)
	c.pending[id] = ch
	payload, _ := json.Marshal(map[string]any{"jsonrpc": "2.0", "id": id, "method": method, "params": params})
	payload = append(payload, '\n')
	_, werr := c.stdin.Write(payload)
	c.mu.Unlock()
	if werr != nil {
		c.mu.Lock()
		delete(c.pending, id)
		c.mu.Unlock()
		return nil, werr
	}
	select {
	case r := <-ch:
		return r.result, r.err
	case <-ctx.Done():
		c.mu.Lock()
		delete(c.pending, id)
		c.mu.Unlock()
		return nil, ctx.Err()
	}
}

// Notify 发无 id 通知。
func (c *AppServerClient) Notify(method string, params any) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.closed {
		return
	}
	payload, _ := json.Marshal(map[string]any{"jsonrpc": "2.0", "method": method, "params": params})
	payload = append(payload, '\n')
	_, _ = c.stdin.Write(payload)
}

// Stream 返回通知 chan（进程退出时关闭）。
func (c *AppServerClient) Stream() <-chan ServerNotification {
	return c.notifyCh
}

// ClearNotifications 丢弃缓冲未消费的通知（预热池取用前清空 warmup 通知）。
func (c *AppServerClient) ClearNotifications() {
	c.mu.Lock()
	defer c.mu.Unlock()
	for {
		select {
		case <-c.notifyCh:
		default:
			return
		}
	}
}

// Close SIGTERM 整组 → grace → SIGKILL；幂等。
func (c *AppServerClient) Close(grace time.Duration) error {
	c.mu.Lock()
	if c.closed {
		c.mu.Unlock()
		return nil
	}
	c.closed = true
	cmd := c.cmd
	c.mu.Unlock()
	if cmd == nil || cmd.Process == nil {
		return nil
	}
	killTree(cmd.Process.Pid, grace)
	return nil
}

func (c *AppServerClient) readLoop(stdout io.Reader) {
	sc := bufio.NewScanner(stdout)
	sc.Buffer(make([]byte, 0, 64*1024), 4*1024*1024) // codex 通知可能很大（diff/命令输出）
	for sc.Scan() {
		line := sc.Bytes()
		if len(line) == 0 {
			continue
		}
		c.handleLine(line)
	}
}

func (c *AppServerClient) handleLine(line []byte) {
	var msg map[string]json.RawMessage
	if err := json.Unmarshal(line, &msg); err != nil {
		core.Warn(context.Background(), "agent", "nonjson", fmt.Sprintf("non-json line: %q", truncStr(string(line), 120)))
		return
	}
	idRaw, hasID := msg["id"]
	hasResult := msg["result"] != nil
	hasError := msg["error"] != nil
	_, hasMethod := msg["method"]

	// response（id + result/error，无 method）
	if hasID && (hasResult || hasError) && !hasMethod {
		var id float64
		_ = json.Unmarshal(idRaw, &id)
		iid := int(id)
		c.mu.Lock()
		ch, ok := c.pending[iid]
		if ok {
			delete(c.pending, iid)
		}
		c.mu.Unlock()
		if !ok {
			return
		}
		if hasError {
			var e struct {
				Message string `json:"message"`
			}
			_ = json.Unmarshal(msg["error"], &e)
			msg := e.Message
			if msg == "" {
				msg = "JSON-RPC error"
			}
			ch <- rpcResult{err: &JsonRpcError{Msg: msg}}
		} else {
			ch <- rpcResult{result: msg["result"]}
		}
		return
	}
	// server-initiated request（id + method）→ 回 method not found（approvalPolicy:never 不应有）
	if hasID && hasMethod {
		var id float64
		_ = json.Unmarshal(idRaw, &id)
		resp, _ := json.Marshal(map[string]any{"jsonrpc": "2.0", "id": int(id), "error": map[string]any{"code": -32601, "message": "not handled"}})
		resp = append(resp, '\n')
		c.mu.Lock()
		_, _ = c.stdin.Write(resp)
		c.mu.Unlock()
		return
	}
	// notification（method，无 id）
	if hasMethod {
		var n ServerNotification
		if err := json.Unmarshal(line, &n); err == nil {
			select {
			case c.notifyCh <- n:
			default:
				// 缓冲满：丢弃（消费者太慢；Phase 1 通知量不会到 1024）。
			}
		}
	}
}

func (c *AppServerClient) handleExit() {
	c.exitMu.Do(func() {
		c.mu.Lock()
		c.exited = true
		c.closed = true
		pending := c.pending
		c.pending = map[int]chan rpcResult{}
		c.mu.Unlock()
		for _, ch := range pending {
			ch <- rpcResult{err: errors.New("app-server exited")}
		}
		close(c.notifyCh)
	})
}

func truncStr(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n]
}

// stderrLogger 把 codex stderr 行 warn 到日志。
type stderrLogger struct{}

func (stderrLogger) Write(p []byte) (int, error) {
	line := truncStr(string(p), 200)
	core.Warn(context.Background(), "agent", "stderr", line)
	return len(p), nil
}
